internals/overlord/logstate/loki.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+type lokiClient struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+}
+
+func newLokiClient(target *plan.LogTarget) logClient {
+	return &lokiClient{
+		target:    target,
+		remoteURL: target.Location,
+		entries:   map[string][]lokiEntry{},
+	}
+}
+
+func (c *lokiClient) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], asLokiEntry(entry))
+	if c.numEntries() >= 10 {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func (c *lokiClient) numEntries() int {
+	sum := 0
+	for _, l := range c.entries {
+		sum += len(l)
+	}
+	return sum
+}
+
+func asLokiEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimRight(entry.Message, "\n"),
+	}
+}
+
+func (c *lokiClient) Flush(ctx context.Context) error {
+	if c.numEntries() == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// Check response status code to see if it was successful
+	if sc := resp.StatusCode; sc < 200 || sc >= 300 {
+		// Request to Loki failed
+		b, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+		if err != nil {
+			b = append(b, []byte("//couldn't read response body: ")...)
+			b = append(b, []byte(err.Error())...)
+		}
+		return fmt.Errorf("cannot send logs to Loki (HTTP %d), response body:\n%s", resp.StatusCode, b)
+	}
+
+	return nil
```

Comment:
I think `io.Copy(io.Discard, resp.Body)` should be safe even if `resp.Body` is huge? Because `io.Discard` is supposed to do nothing. I'm having trouble finding what I want in the Go source code (too many levels of abstraction), but I might try to test it myself.

Reply:
Well, that means if the server is (buggily) sending a huge 1GB response, we'll read it all. Whereas if we limit it, we'll only read up to max bytes before closing the connection, which will forcibly terminate the TCP connection in the case of a buggily huge response. Edge case of an edge case, I know...

Reply:
nope, did some testing and copying to `io.Discard` is still slow
https://go.dev/play/p/whIS19engLK
(you have to run this locally, it doesn't work on Go Playground)

Copying 500MB to `io.Discard` took 0.4 seconds on my machine, and it seems to scale linearly.

So I'm happy to add `io.Copy(io.Discard, io.LimitReader(resp.Body, 1024))` to enable connection reuse, but probably we should make the limit larger - 1KB is not a lot. I think we can eat 1MB in under a millisecond, so that shouldn't be a hold up.